File: src/jsonrpc/eth/call/eth_call.go
package eth

import (
	"encoding/json"
	"fmt"

	"github.com/ethereum/execution-apis/types"
)

// Executes a new message call immediately without creating a transaction on the block chain.
//
// Example:
// Transaction: ...
// Block: "latest"
// Result: "0x"
//
// Implements the eth_call JSON-RPC method.

// Method is the JSON-RPC method name
const Method = "eth_call"

// Params represents the parameters for eth_call
type Params struct {
	// Transaction object generic to all types
	Transaction types.Quantity `json:"-"`
	// Block number, tag, or block hash
	Block types.BlockSpec `json:"-"`
}

// MarshalJSON implements json.Marshaler for Params.
// JSON-RPC 2.0 uses positional array parameters.
func (p Params) MarshalJSON() ([]byte, error) {
	return json.Marshal([]interface{}{
		p.Transaction,
		p.Block,
	})
}

// UnmarshalJSON implements json.Unmarshaler for Params.
func (p *Params) UnmarshalJSON(data []byte) error {
	var arr []json.RawMessage
	if err := json.Unmarshal(data, &arr); err != nil {
		return err
	}

	if len(arr) != 2 {
		return fmt.Errorf("expected 2 parameters, got %d", len(arr))
	}

	if err := json.Unmarshal(arr[0], &p.Transaction); err != nil {
		return fmt.Errorf("parameter 0 (Transaction): %w", err)
	}

	if err := json.Unmarshal(arr[1], &p.Block); err != nil {
		return fmt.Errorf("parameter 1 (Block): %w", err)
	}

	return nil
}

// Result represents the result for eth_call
//
// hex encoded bytes
type Result struct {
	Value types.Quantity `json:"-"`
}

// MarshalJSON implements json.Marshaler for Result.
func (r Result) MarshalJSON() ([]byte, error) {
	return json.Marshal(r.Value)
}

// UnmarshalJSON implements json.Unmarshaler for Result.
func (r *Result) UnmarshalJSON(data []byte) error {
	return json.Unmarshal(data, &r.Value)
}

File: src/jsonrpc/eth/call/eth_call.js
/**
 * @fileoverview eth_call JSON-RPC method
 */

/**
 * @typedef {import('../../../primitives/Address/AddressType.js').AddressType} Address
 * @typedef {import('../../types/index.js').Hash} Hash
 * @typedef {import('../../types/index.js').Quantity} Quantity
 * @typedef {import('../../types/index.js').BlockTag} BlockTag
 * @typedef {import('../../types/index.js').BlockSpec} BlockSpec
 * @typedef {import('../../../provider/types.js').RequestArguments} RequestArguments
 */

/**
 * Transaction call parameters
 *
 * @typedef {object} CallParams
 * @property {Address} [from] - Transaction sender
 * @property {Address} to - Transaction recipient
 * @property {Quantity} [gas] - Gas provided for transaction
 * @property {Quantity} [gasPrice] - Gas price
 * @property {Quantity} [value] - Value sent with transaction
 * @property {`0x${string}`} [data] - Call data
 */

/**
 * Executes a new message call immediately without creating a transaction on the block chain.
 *
 * @example
 * Transaction: ...
 * Block: "latest"
 * Result: "0x"
 *
 * Implements the `eth_call` JSON-RPC method.
 */

/** The JSON-RPC method name */
export const method = "eth_call";

/**
 * Result for `eth_call`
 *
 * hex encoded bytes
 *
 * @typedef {Quantity} Result
 */

/**
 * Creates an eth_call JSON-RPC request
 *
 * @param {CallParams} params - Transaction call parameters
 * @param {BlockSpec} [block='latest'] - Block number, tag, or block hash
 * @returns {RequestArguments}
 *
 * @example
 * ```js
 * const request = CallRequest({
 *   to: '0x742d35Cc6634C0532925a3b844Bc9e7595f0bEb0',
 *   data: '0x...'
 * }, 'latest')
 * ```
 */
export function CallRequest(params, block = "latest") {
	return { method, params: [params, block] };
}
